core/src/commonMain/kotlin/com/powersync/attachments/AttachmentQueue.kt

```diff
@@ -0,0 +1,489 @@
+    /**
+     * Optional meta data for the attachment record
+     */
+    public val metaData: String? = null,
+) {
+    init {
+        require(fileExtension != null || filename != null) {
+            "Either fileExtension or filename must be provided."
+        }
+    }
+}
+
+/**
+ * Class used to implement the attachment queue
+ * Requires a PowerSyncDatabase, an implementation of
+ * AbstractRemoteStorageAdapter and an attachment directory name which will
+ * determine which folder attachments are stored into.
+ */
+public open class AttachmentQueue(
+    /**
+     * PowerSync database client
+     */
+    public val db: PowerSyncDatabase,
+    /**
+     * Adapter which interfaces with the remote storage backend
+     */
+    public val remoteStorage: RemoteStorage,
+    /**
+     * Directory name where attachment files will be written to disk.
+     * This will be created if it does not exist
+     */
+    private val attachmentsDirectory: String,
+    /**
+     * A flow for the current state of local attachments
+     * ```kotlin
+     *     watchedAttachment = db.watch(
+     *             sql =
+     *                 """
+     *                 SELECT
+     *                     photo_id as id
+     *                 FROM
+     *                     checklists
+     *                 WHERE
+     *                     photo_id IS NOT NULL
+     *                 """,
+     *         ) { cursor ->
+     *             WatchedAttachmentItem(
+     *                 id = cursor.getString("id"),
+     *                 fileExtension = "jpg",
+     *             )
+     *         }
+     * ```
+     */
+    private val watchedAttachments: Flow<List<WatchedAttachmentItem>>,
+    /**
+     * Provides access to local filesystem storage methods
+     */
+    public val localStorage: LocalStorage = IOLocalStorageAdapter(),
+    /**
+     * SQLite table where attachment state will be recorded
+     */
+    private val attachmentsQueueTableName: String = DEFAULT_TABLE_NAME,
+    /**
+     * Attachment operation error handler. This specified if failed attachment operations
+     * should be retried.
+     */
+    private val errorHandler: SyncErrorHandler? = null,
+    /**
+     * Periodic interval to trigger attachment sync operations
+     */
+    private val syncInterval: Duration = 30.seconds,
+    /**
+     * Archived attachments can be used as a cache which can be restored if an attachment id
+     * reappears after being removed. This parameter defines how many archived records are retained.
+     * Records are deleted once the number of items exceeds this value.
+     */
+    private val archivedCacheLimit: Long = 100,
+    /**
+     * Throttles remote sync operations triggering
+     */
+    private val syncThrottleDuration: Duration = 1.seconds,
+    /**
+     * Creates a list of subdirectories in the {attachmentDirectoryName} directory
+     */
+    private val subdirectories: List<String>? = null,
```

Comment:
I remember a FlutterFlow user asking for dynamic subdirectories (because they have something like a `projects` table and want to store images in `$project/$image.png`. Users can be added and removed from projects dynamically).

So I'm just wondering if it might be easier to just support slashes as path separators (creating the directory automatically) instead of listing subdirectories statically. But consistency with our other SDKs is also important, so I'm not sure if it's worth it.

Reply:
I suppose a dev could override the local storage implementation to upsert directories if they required that functionality.